{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Nathan removed functions related to terminal commands from given library\n",
    "from part_ncbiblast_library import *\n",
    "import json\n",
    "import pandas as pd\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.common.action_chains import ActionChains\n",
    "from selenium.webdriver.chrome.options import Options\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.chrome.service import Service as ChromeService\n",
    "from webdriver_manager.chrome import ChromeDriverManager\n",
    "import os, sys\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Getting sequence to BLAST from user, optional email and name of BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test seq: MIDNAYHIDS (4 hits Uniprot, 0 hits Phytozome)\n",
    "#test seq 2: MARPFTLSLLSLCLLLSAWSCFGGSSSTNRFNICQLNSLNALKPDHRVETDGGLVETWSSRHPELECAGVTVTRRTLYRNGFQMPSYSPYSQMIMAIQGKGALGLALSGCAETYEEPAKESSSSSQKPSDSHQKIRQFDQGHVMLIPRGVPFWIFNTGDEPLITVTLLDTSSEDNQLDQSPREFYLAGNPDIEHPEAMKEKQQQQAEEEGGNVLSGFGKRFLARALNIDQDTANKLISPDDEMKQIVKLKEGLSVISPKWQGQQEDEDEDDDDEDEDESVSRPSRRPSHGKRVHKEEETVVEPYPHGKHVHKEVEKEVEPLPPRKHVHKEEEKEIEPLPPRRSRHHHDEGEDEGEEEEKPRARRTRGPTPSPKGEGHRGVEEEDESEDTKGHKTRHEKTRHEKSWKEHRPEGEDVEKGEAHEEWETRPSKDKPHGSNGLDETICSSKLQFNIARPKGADFYNPKAGRIKNLNSQSLPALQHFGLSAQYVVLYKNGIYSPHWNMDANSVIYVIRGQGQVRVVNNEGIVMFDDELKKGQLLVVPQNFMVAEEAGDQGFEYVVFKTNDNAVTSYLKETFRAFPAEVLVNIYKLKHSQVHDLKYNGNLGPLVNPENSLDQSS\n",
    "#(250 hits Uniprot, 100 hits Phytozome)\n",
    "seq = str(input(\"Protein AA Sequence to Compare: \"))\n",
    "email = (input(\"What's your email: \") or 'tafsia.shikdar@beyondmeat.com')\n",
    "title = str(input('Name this run: ') or 'Protein Search')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Setting params for Uniprot API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Required for uniprot blast. Only input params are email and sequence, everything else is default\n",
    "params = {\n",
    "        \"program\":\"blastp\",\n",
    "        \"matrix\":\"PAM30\",\n",
    "        \"alignments\":\"250\",\n",
    "        \"scores\":\"250\",\n",
    "        \"exp\":\"10\",\n",
    "        \"filter\":\"F\",\n",
    "        \"gapalign\":\"true\",\n",
    "        \"compstats\":\"F\",\n",
    "        \"align\":\"0\",\n",
    "        \"stype\":\"protein\",\n",
    "        \"sequence\": seq,\n",
    "        \"database\":\"uniprotkb_refprotswissprot\"\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running Uniprot BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Running Uniprot job after asking for a name (otherwise default)\n",
    "jobId = serviceRun(email, title, params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Starting Phytozome webscraping/BLAST simultaneously to save time; setting up remotely controlled browser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "#object of Options class\n",
    "op = webdriver.ChromeOptions()\n",
    "#browser preferences i.e. saving all downloaded files into certain location to find later\n",
    "p = {'download.default_directory':'C:\\\\Users\\\\Tafsia Shikdar\\\\Downloads\\\\phytozome_download'}\n",
    "#add options to browser i.e. where to save downloaded files, and not opening a new window visibly\n",
    "op.add_experimental_option('prefs', p)\n",
    "#op.add_argument('headless')\n",
    "#webdriver (interface for remotely controlling browsers) location, specifically for Google Chrome\n",
    "service = ChromeService(executable_path='C:/Users/Tafsia Shikdar/Desktop/protein-uniprot-blast/chromedriver.exe')\n",
    "driver = webdriver.Chrome(service=service, options=op)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running Phytozome BLAST (using Selenium)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load phytozome blast webpage\n",
    "driver.get(\"https://phytozome-next.jgi.doe.gov/blast-search\")\n",
    "#HTML code showing location of search text box, and clicking into text box\n",
    "click = WebDriverWait(driver, 1000).until(EC.presence_of_element_located((By.XPATH, '/html/body/div/div/div/main/div/div/div/div/div/div[1]/div[2]/div/ol/li/span/label/span')))\n",
    "driver.execute_script(\"arguments[0].click();\", click)\n",
    "#input sequence into box\n",
    "search = WebDriverWait(driver, 1000).until(EC.presence_of_element_located((By.XPATH, '/html/body/div/div/div/main/div/div/div/div/form/fieldset/textarea')))\n",
    "driver.execute_script(\"arguments[0].value= '\" + seq + \"'\", search)\n",
    "#had some issues with inputting sequence - worked after adding and deleting a space after seq is input\n",
    "search.send_keys(Keys.SPACE)\n",
    "search.send_keys(Keys.BACK_SPACE)\n",
    "#HTML code showing location of enter/go button\n",
    "go = WebDriverWait(driver, 1000).until(EC.presence_of_element_located((By.XPATH, '/html/body/div/div/div/main/div/div/div/div/form/fieldset/div/ul/li[1]/span')))\n",
    "driver.execute_script(\"arguments[0].click();\", go)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Waiting for Uniprot BLAST to complete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n",
      "RUNNING\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\Tafsia Shikdar\\Desktop\\protein-uniprot-blast\\overall.ipynb Cell 13\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Tafsia%20Shikdar/Desktop/protein-uniprot-blast/overall.ipynb#X11sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39m#This function checks Uniprot job status every 3 seconds until it is done\u001b[39;00m\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/Tafsia%20Shikdar/Desktop/protein-uniprot-blast/overall.ipynb#X11sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m clientPoll(jobId)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Tafsia%20Shikdar/Desktop/protein-uniprot-blast/overall.ipynb#X11sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39m#Retrieving Uniprot result as a json file and loading it in so it can be parsed as dictionary\u001b[39;00m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Tafsia%20Shikdar/Desktop/protein-uniprot-blast/overall.ipynb#X11sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m file \u001b[39m=\u001b[39m serviceGetResult(jobId, \u001b[39m'\u001b[39m\u001b[39mjson\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\Tafsia Shikdar\\Desktop\\protein-uniprot-blast\\part_ncbiblast_library.py:327\u001b[0m, in \u001b[0;36mclientPoll\u001b[1;34m(jobId)\u001b[0m\n\u001b[0;32m    325\u001b[0m         \u001b[39mprint\u001b[39m(result)\n\u001b[0;32m    326\u001b[0m     \u001b[39mif\u001b[39;00m result \u001b[39m==\u001b[39m \u001b[39mu\u001b[39m\u001b[39m'\u001b[39m\u001b[39mRUNNING\u001b[39m\u001b[39m'\u001b[39m \u001b[39mor\u001b[39;00m result \u001b[39m==\u001b[39m \u001b[39mu\u001b[39m\u001b[39m'\u001b[39m\u001b[39mPENDING\u001b[39m\u001b[39m'\u001b[39m:\n\u001b[1;32m--> 327\u001b[0m         time\u001b[39m.\u001b[39;49msleep(pollFreq)\n\u001b[0;32m    328\u001b[0m printDebugMessage(\u001b[39mu\u001b[39m\u001b[39m'\u001b[39m\u001b[39mclientPoll\u001b[39m\u001b[39m'\u001b[39m, \u001b[39mu\u001b[39m\u001b[39m'\u001b[39m\u001b[39mEnd\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m1\u001b[39m)\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#This function checks Uniprot job status every 3 seconds until it is done\n",
    "clientPoll(jobId)\n",
    "#Retrieving Uniprot result as a json file and loading it in so it can be parsed as dictionary\n",
    "file = serviceGetResult(jobId, 'json')\n",
    "results = json.loads(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Chains are split into alpha and beta chains according to Ryan Prestangen's algorithm to split chains into alpha and beta (assumed for 11S globulin proteins specifically). Let me know if this needs to be updated - in future we can look at machine learning models to replace this manually designed splitting method.\n",
    "Algorithm: Will not split proteins smaller than 20 amino acids, or if there are no C (cysteine) residues past the first 5 residues in the chain.\n",
    "First, tries to find an NG terminus (asparagine immediately followed by a threonine) with a C 7 residues after it. If this is found, chain is split 5 amino acids before the C.\n",
    "Secondly, if such a spot cannot be found, count the number of cysteines in the SECOND HALF of the peptide chain. IF there are more than 2 cysteines in the second half, chain is split 5 spots before the third to last cysteine in the SECOND HALF of protein. Otherwise, if there are only 1 or 2 cysteines, chain is split 5 spots before the first cysteine in the SECOND HALF of protein.\n",
    "Finally, if there are no cysteines in the second half of the sequence, split chain five spots before the last cysteine in the FIRST HALF of the protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def splitting_chains(protein_seq):\n",
    "    '''Splits sequence into two chains.\n",
    "    Parameters: (str) AA sequence\n",
    "    Returns: (tuple) ((str) Method of Split, \n",
    "                      (int) Index for start of second chain OR (str) Reason for not splitting protein)'''\n",
    "    #will only do this if chain is longer than 20 AA\n",
    "    size = len(protein_seq)\n",
    "    if size < 20:\n",
    "        return ('No Split', 'Protein Too Small to Split')\n",
    "    ng_indexes = []\n",
    "    c_indexes = []\n",
    "    for i in range(size):\n",
    "        aa = protein_seq[i]\n",
    "        #cysteines in the first 5 AAs not counted\n",
    "        if i > 5 and aa == 'C':\n",
    "            c_indexes.append(i)\n",
    "        #First method: trying to find a cysteine 7 spots after an NG. If this is found, chain should be \n",
    "        #split 2 spots after the NG index/5 spots before the cysteine\n",
    "        try:\n",
    "            if aa == 'N' and protein_seq[i+1] == 'G':\n",
    "                ng_indexes.append(i)\n",
    "                if protein_seq[i+7] == 'C':\n",
    "                    return ('NG Terminus at ' + str(i), i+2)\n",
    "        except:\n",
    "            continue\n",
    "    #Second method: Find all cysteines in second half of sequence, and base split on number of Cs found\n",
    "    #If no Cs, can't split\n",
    "    c_count = len(c_indexes)\n",
    "    if c_count == 0:\n",
    "        return ('No Split', 'Protein Has No Viable Cysteines')\n",
    "    half = size/2\n",
    "    #Only using the c_indexes found in second half of protein sequence\n",
    "    i = 0\n",
    "    while i < c_count and c_indexes[i] < half:\n",
    "        i += 1\n",
    "    second_half_c = c_indexes[i+1:]\n",
    "    second_c_count = len(second_half_c)\n",
    "    if second_c_count != 0:\n",
    "        #If more than 2 cystines in second half of protein seq, chain should be split 5 spots \n",
    "        #before the third to last cysteine\n",
    "        if second_c_count > 2:\n",
    "            c_loc = second_half_c[-3]\n",
    "            return ('Third to Last Cysteine in Second Half at ' + str(c_loc), c_loc - 5)\n",
    "        #If only 1 or 2 cystines in second half of protein seq, chain should be split 5 spots before the first\n",
    "        #cysteine.\n",
    "        else:\n",
    "            c_loc = second_half_c[0]\n",
    "            return ('First Cysteine in Second Half at ' + str(c_loc), c_loc - 5)\n",
    "    #Third Method: Split chain 5 spots before last cysteine in first half of protein seq\n",
    "    c_loc = c_indexes[:i][-1] \n",
    "    return ('Last Cysteine in First Half at ' + str(c_loc), c_loc - 5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Results:250\n"
     ]
    }
   ],
   "source": [
    "data = {\n",
    "    'Website': [],\n",
    "    'Link':[], \n",
    "    'Entry': [], \n",
    "    'Entry Name': [], \n",
    "    'Protein Name': [], \n",
    "    'Organism': [], \n",
    "    'Gene Name': [], \n",
    "    'Length': [], \n",
    "    'Sequence': [], \n",
    "    'Percent Identity': [],\n",
    "    'Split Method': [],\n",
    "    'Chain 1 Sequence': [],\n",
    "    'Chain 2 Sequence': []\n",
    "    }\n",
    "\n",
    "#Creating function to fill in sequence split columns using given sequence and chain_splitting algorithm\n",
    "def split_chain_func(chain_sequence):\n",
    "    split_chain_method, split_chain_result = splitting_chains(chain_sequence)\n",
    "    data['Split Method'].append(split_chain_method)\n",
    "    if type(split_chain_result) != int:\n",
    "        data['Chain 1 Sequence'].append(split_chain_result)\n",
    "        data['Chain 2 Sequence'].append(split_chain_result)\n",
    "    else:\n",
    "        data['Chain 1 Sequence'].append(chain_sequence[:split_chain_result])\n",
    "        data['Chain 2 Sequence'].append(chain_sequence[split_chain_result:])  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Searching through Uniprot JSON, pulling the data needed into a dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uniprot1\n",
      "Uniprot2\n",
      "Uniprot3\n",
      "Uniprot4\n",
      "Uniprot5\n",
      "Uniprot6\n",
      "Uniprot7\n",
      "Uniprot8\n",
      "Uniprot9\n",
      "Uniprot10\n",
      "Uniprot11\n",
      "Uniprot12\n",
      "Uniprot13\n",
      "Uniprot14\n",
      "Uniprot15\n",
      "Uniprot16\n",
      "Uniprot17\n",
      "Uniprot18\n",
      "Uniprot19\n",
      "Uniprot20\n",
      "Uniprot21\n",
      "Uniprot22\n",
      "Uniprot23\n",
      "Uniprot24\n",
      "Uniprot25\n",
      "Uniprot26\n",
      "Uniprot27\n",
      "Uniprot28\n",
      "Uniprot29\n",
      "Uniprot30\n",
      "Uniprot31\n",
      "Uniprot32\n",
      "Uniprot33\n",
      "Uniprot34\n",
      "Uniprot35\n",
      "Uniprot36\n",
      "Uniprot37\n",
      "Uniprot38\n",
      "Uniprot39\n",
      "Uniprot40\n",
      "Uniprot41\n",
      "Uniprot42\n",
      "Uniprot43\n",
      "Uniprot44\n",
      "Uniprot45\n",
      "Uniprot46\n",
      "Uniprot47\n",
      "Uniprot48\n",
      "Uniprot49\n",
      "Uniprot50\n",
      "Uniprot51\n",
      "Uniprot52\n",
      "Uniprot53\n",
      "Uniprot54\n",
      "Uniprot55\n",
      "Uniprot56\n",
      "Uniprot57\n",
      "Uniprot58\n",
      "Uniprot59\n",
      "Uniprot60\n",
      "Uniprot61\n",
      "Uniprot62\n",
      "Uniprot63\n",
      "Uniprot64\n",
      "Uniprot65\n",
      "Uniprot66\n",
      "Uniprot67\n",
      "Uniprot68\n",
      "Uniprot69\n",
      "Uniprot70\n",
      "Uniprot71\n",
      "Uniprot72\n",
      "Uniprot73\n",
      "Uniprot74\n",
      "Uniprot75\n",
      "Uniprot76\n",
      "Uniprot77\n",
      "Uniprot78\n",
      "Uniprot79\n",
      "Uniprot80\n",
      "Uniprot81\n",
      "Uniprot82\n",
      "Uniprot83\n",
      "Uniprot84\n",
      "Uniprot85\n",
      "Uniprot86\n",
      "Uniprot87\n",
      "Uniprot88\n",
      "Uniprot89\n",
      "Uniprot90\n",
      "Uniprot91\n",
      "Uniprot92\n",
      "Uniprot93\n",
      "Uniprot94\n",
      "Uniprot95\n",
      "Uniprot96\n",
      "Uniprot97\n",
      "Uniprot98\n",
      "Uniprot99\n",
      "Uniprot100\n",
      "Uniprot101\n",
      "Uniprot102\n",
      "Uniprot103\n",
      "Uniprot104\n",
      "Uniprot105\n",
      "Uniprot106\n",
      "Uniprot107\n",
      "Uniprot108\n",
      "Uniprot109\n",
      "Uniprot110\n",
      "Uniprot111\n",
      "Uniprot112\n",
      "Uniprot113\n",
      "Uniprot114\n",
      "Uniprot115\n",
      "Uniprot116\n",
      "Uniprot117\n",
      "Uniprot118\n",
      "Uniprot119\n",
      "Uniprot120\n",
      "Uniprot121\n",
      "Uniprot122\n",
      "Uniprot123\n",
      "Uniprot124\n",
      "Uniprot125\n",
      "Uniprot126\n",
      "Uniprot127\n",
      "Uniprot128\n",
      "Uniprot129\n",
      "Uniprot130\n",
      "Uniprot131\n",
      "Uniprot132\n",
      "Uniprot133\n",
      "Uniprot134\n",
      "Uniprot135\n",
      "Uniprot136\n",
      "Uniprot137\n",
      "Uniprot138\n",
      "Uniprot139\n",
      "Uniprot140\n",
      "Uniprot141\n",
      "Uniprot142\n",
      "Uniprot143\n",
      "Uniprot144\n",
      "Uniprot145\n",
      "Uniprot146\n",
      "Uniprot147\n",
      "Uniprot148\n",
      "Uniprot149\n",
      "Uniprot150\n",
      "Uniprot151\n",
      "Uniprot152\n",
      "Uniprot153\n",
      "Uniprot154\n",
      "Uniprot155\n",
      "Uniprot156\n",
      "Uniprot157\n",
      "Uniprot158\n",
      "Uniprot159\n",
      "Uniprot160\n",
      "Uniprot161\n",
      "Uniprot162\n",
      "Uniprot163\n",
      "Uniprot164\n",
      "Uniprot165\n",
      "Uniprot166\n",
      "Uniprot167\n",
      "Uniprot168\n",
      "Uniprot169\n",
      "Uniprot170\n",
      "Uniprot171\n",
      "Uniprot172\n",
      "Uniprot173\n",
      "Uniprot174\n",
      "Uniprot175\n",
      "Uniprot176\n",
      "Uniprot177\n",
      "Uniprot178\n",
      "Uniprot179\n",
      "Uniprot180\n",
      "Uniprot181\n",
      "Uniprot182\n",
      "Uniprot183\n",
      "Uniprot184\n",
      "Uniprot185\n",
      "Uniprot186\n",
      "Uniprot187\n",
      "Uniprot188\n",
      "Uniprot189\n",
      "Uniprot190\n",
      "Uniprot191\n",
      "Uniprot192\n",
      "Uniprot193\n",
      "Uniprot194\n",
      "Uniprot195\n",
      "Uniprot196\n",
      "Uniprot197\n",
      "Uniprot198\n",
      "Uniprot199\n",
      "Uniprot200\n",
      "Uniprot201\n",
      "Uniprot202\n",
      "Uniprot203\n",
      "Uniprot204\n",
      "Uniprot205\n",
      "Uniprot206\n",
      "Uniprot207\n",
      "Uniprot208\n",
      "Uniprot209\n",
      "Uniprot210\n",
      "Uniprot211\n",
      "Uniprot212\n",
      "Uniprot213\n",
      "Uniprot214\n",
      "Uniprot215\n",
      "Uniprot216\n",
      "Uniprot217\n",
      "Uniprot218\n",
      "Uniprot219\n",
      "Uniprot220\n",
      "Uniprot221\n",
      "Uniprot222\n",
      "Uniprot223\n",
      "Uniprot224\n",
      "Uniprot225\n",
      "Uniprot226\n",
      "Uniprot227\n",
      "Uniprot228\n",
      "Uniprot229\n",
      "Uniprot230\n",
      "Uniprot231\n",
      "Uniprot232\n",
      "Uniprot233\n",
      "Uniprot234\n",
      "Uniprot235\n",
      "Uniprot236\n",
      "Uniprot237\n",
      "Uniprot238\n",
      "Uniprot239\n",
      "Uniprot240\n",
      "Uniprot241\n",
      "Uniprot242\n",
      "Uniprot243\n",
      "Uniprot244\n",
      "Uniprot245\n",
      "Uniprot246\n",
      "Uniprot247\n",
      "Uniprot248\n",
      "Uniprot249\n",
      "Uniprot250\n"
     ]
    }
   ],
   "source": [
    "#Manually searched through retrieved json file to identify keys of the info we needed - may be an easier/less hard coded way to do this via API queries\n",
    "columns_to_keys = {\n",
    "    'Link':'hit_url', \n",
    "    'Entry': 'hit_acc', \n",
    "    'Entry Name': 'hit_id', \n",
    "    'Protein Name': 'hit_uni_de', \n",
    "    'Organism': 'hit_uni_os', \n",
    "    'Gene Name': 'hit_uni_gn', \n",
    "    'Length': 'hit_len',\n",
    "}\n",
    "\n",
    "#in case info doesn't exist for entry, default message fills the dataframe\n",
    "def add_info(column):\n",
    "    result_key = columns_to_keys[column]\n",
    "    if result_key in hit:\n",
    "        data[column].append(hit[result_key])\n",
    "    else:\n",
    "        data[column].append('No Data Found')\n",
    "\n",
    "#while building df, will let user know total number of entries and progress\n",
    "print('Total Uniprot Results:' + str(len(results['hits'])))\n",
    "#goes through each hit entry to pull the necessary info using manually matched keys to data desired\n",
    "for hit in results['hits']:\n",
    "    print('Uniprot' + str(hit['hit_num']))\n",
    "    data['Website'].append('Uniprot')\n",
    "    for c in columns_to_keys:\n",
    "        add_info(c)\n",
    "    #adding percent identity, unlike other desired data, this one was nested in the original json file\n",
    "    try:\n",
    "        data['Percent Identity'].append(hit['hit_hsps'][0]['hsp_identity'])\n",
    "    except:\n",
    "        data['Percent Identity'].append('No Data Found')\n",
    "    #sequence not in original data file, so querying api for specific entry to get that data\n",
    "    url = 'https://rest.uniprot.org/uniprotkb/' + hit['hit_acc'] + '.json'\n",
    "    info = json.loads(restRequest(url))\n",
    "    full_seq = info['sequence']['value'].upper()\n",
    "    data['Sequence'].append(full_seq)\n",
    "    #Splitting chain\n",
    "    split_chain_func(full_seq) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Moving onto Phytozome Blast...first downloading blast results file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['blastp_job_616585 (1).tsv', 'blastp_job_616585.tsv', 'blastp_job_633883.tsv', 'blastp_job_633957.tsv', 'blastp_job_635781.tsv', 'blastp_job_639705.tsv', 'blastp_job_639805.tsv', 'blastp_job_639889.tsv', 'blastp_job_639912.tsv', 'blastp_job_639941.tsv', 'blastp_job_640337 (1).tsv', 'blastp_job_640337.tsv', 'blastp_job_641493.tsv', 'blastp_job_641530.tsv', 'blastp_job_644826.tsv', 'd10e545c-0055-448a-b858-3cb6e7d8f71c.tmp', 'export.csv', 'GO_terms_for_proteins.json', 'services.json', 'VuZN016.07G092700.functionalAnnotations.tsv']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Views</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Species</th>\n",
       "      <th>E-value</th>\n",
       "      <th>% identity</th>\n",
       "      <th>Align len</th>\n",
       "      <th>Strands</th>\n",
       "      <th>Query ID</th>\n",
       "      <th>Query from</th>\n",
       "      <th>Query to</th>\n",
       "      <th>Target from</th>\n",
       "      <th>Target to</th>\n",
       "      <th>Bitscore</th>\n",
       "      <th># identical</th>\n",
       "      <th>Positives</th>\n",
       "      <th>Gaps</th>\n",
       "      <th>Query len</th>\n",
       "      <th>Target len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>phytozome-next.jgi.doe.gov/report/protein/Vung...</td>\n",
       "      <td>VuZN016.07G092700.1.p</td>\n",
       "      <td>V.unguiculata_ZN016 v1.2</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>85</td>\n",
       "      <td>620</td>\n",
       "      <td>+/+</td>\n",
       "      <td>seq.submitted 2022-08-22T18:41:21.472Z UTC</td>\n",
       "      <td>1</td>\n",
       "      <td>618</td>\n",
       "      <td>1</td>\n",
       "      <td>611</td>\n",
       "      <td>877.085</td>\n",
       "      <td>527</td>\n",
       "      <td>567</td>\n",
       "      <td>11</td>\n",
       "      <td>618</td>\n",
       "      <td>612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>phytozome-next.jgi.doe.gov/report/protein/Vung...</td>\n",
       "      <td>VuTZ30.07G092400.1.p</td>\n",
       "      <td>V.unguiculata_TZ30 v1.2</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>85</td>\n",
       "      <td>620</td>\n",
       "      <td>+/+</td>\n",
       "      <td>seq.submitted 2022-08-22T18:41:21.472Z UTC</td>\n",
       "      <td>1</td>\n",
       "      <td>618</td>\n",
       "      <td>1</td>\n",
       "      <td>611</td>\n",
       "      <td>877.085</td>\n",
       "      <td>527</td>\n",
       "      <td>567</td>\n",
       "      <td>11</td>\n",
       "      <td>618</td>\n",
       "      <td>612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>phytozome-next.jgi.doe.gov/report/protein/Vung...</td>\n",
       "      <td>VuCB5-2.07G096300.1.p</td>\n",
       "      <td>V.unguiculata_CB5-2 v1.1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>85</td>\n",
       "      <td>620</td>\n",
       "      <td>+/+</td>\n",
       "      <td>seq.submitted 2022-08-22T18:41:21.472Z UTC</td>\n",
       "      <td>1</td>\n",
       "      <td>618</td>\n",
       "      <td>1</td>\n",
       "      <td>611</td>\n",
       "      <td>877.085</td>\n",
       "      <td>527</td>\n",
       "      <td>567</td>\n",
       "      <td>11</td>\n",
       "      <td>618</td>\n",
       "      <td>612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>phytozome-next.jgi.doe.gov/report/protein/Vung...</td>\n",
       "      <td>VuSanzi.07G093900.1.p</td>\n",
       "      <td>V.unguiculata_Sanzi v1.1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>85</td>\n",
       "      <td>620</td>\n",
       "      <td>+/+</td>\n",
       "      <td>seq.submitted 2022-08-22T18:41:21.472Z UTC</td>\n",
       "      <td>1</td>\n",
       "      <td>618</td>\n",
       "      <td>1</td>\n",
       "      <td>611</td>\n",
       "      <td>877.085</td>\n",
       "      <td>527</td>\n",
       "      <td>567</td>\n",
       "      <td>11</td>\n",
       "      <td>618</td>\n",
       "      <td>612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>phytozome-next.jgi.doe.gov/report/protein/Vung...</td>\n",
       "      <td>VuUCR779.07G095400.1.p</td>\n",
       "      <td>V.unguiculata_UCR779 v1.1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>85</td>\n",
       "      <td>620</td>\n",
       "      <td>+/+</td>\n",
       "      <td>seq.submitted 2022-08-22T18:41:21.472Z UTC</td>\n",
       "      <td>1</td>\n",
       "      <td>618</td>\n",
       "      <td>1</td>\n",
       "      <td>611</td>\n",
       "      <td>877.085</td>\n",
       "      <td>527</td>\n",
       "      <td>567</td>\n",
       "      <td>11</td>\n",
       "      <td>618</td>\n",
       "      <td>612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>phytozome-next.jgi.doe.gov/report/protein/Mgut...</td>\n",
       "      <td>MgNONTOL.G0505.1.p</td>\n",
       "      <td>M.guttatus_NONTOL v3.1</td>\n",
       "      <td>6.440000e-34</td>\n",
       "      <td>36</td>\n",
       "      <td>222</td>\n",
       "      <td>+/+</td>\n",
       "      <td>seq.submitted 2022-08-22T18:41:21.472Z UTC</td>\n",
       "      <td>40</td>\n",
       "      <td>258</td>\n",
       "      <td>73</td>\n",
       "      <td>273</td>\n",
       "      <td>143.280</td>\n",
       "      <td>80</td>\n",
       "      <td>124</td>\n",
       "      <td>24</td>\n",
       "      <td>618</td>\n",
       "      <td>495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>phytozome-next.jgi.doe.gov/report/protein/Acoe...</td>\n",
       "      <td>Aqcoe5G027700.1.p</td>\n",
       "      <td>A.coerulea v3.1</td>\n",
       "      <td>3.310000e-48</td>\n",
       "      <td>46</td>\n",
       "      <td>184</td>\n",
       "      <td>+/+</td>\n",
       "      <td>seq.submitted 2022-08-22T18:41:21.472Z UTC</td>\n",
       "      <td>430</td>\n",
       "      <td>610</td>\n",
       "      <td>288</td>\n",
       "      <td>471</td>\n",
       "      <td>182.956</td>\n",
       "      <td>84</td>\n",
       "      <td>126</td>\n",
       "      <td>3</td>\n",
       "      <td>618</td>\n",
       "      <td>482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>phytozome-next.jgi.doe.gov/report/protein/Acoe...</td>\n",
       "      <td>Aqcoe5G027700.1.p</td>\n",
       "      <td>A.coerulea v3.1</td>\n",
       "      <td>9.160000e-36</td>\n",
       "      <td>35</td>\n",
       "      <td>243</td>\n",
       "      <td>+/+</td>\n",
       "      <td>seq.submitted 2022-08-22T18:41:21.472Z UTC</td>\n",
       "      <td>34</td>\n",
       "      <td>258</td>\n",
       "      <td>32</td>\n",
       "      <td>274</td>\n",
       "      <td>148.288</td>\n",
       "      <td>86</td>\n",
       "      <td>130</td>\n",
       "      <td>18</td>\n",
       "      <td>618</td>\n",
       "      <td>482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>phytozome-next.jgi.doe.gov/report/protein/Pami...</td>\n",
       "      <td>FUN_051248-T1</td>\n",
       "      <td>P.amilis v1.0</td>\n",
       "      <td>3.590000e-48</td>\n",
       "      <td>45</td>\n",
       "      <td>190</td>\n",
       "      <td>+/+</td>\n",
       "      <td>seq.submitted 2022-08-22T18:41:21.472Z UTC</td>\n",
       "      <td>415</td>\n",
       "      <td>601</td>\n",
       "      <td>274</td>\n",
       "      <td>463</td>\n",
       "      <td>182.956</td>\n",
       "      <td>86</td>\n",
       "      <td>127</td>\n",
       "      <td>3</td>\n",
       "      <td>618</td>\n",
       "      <td>487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>phytozome-next.jgi.doe.gov/report/protein/Pami...</td>\n",
       "      <td>FUN_051248-T1</td>\n",
       "      <td>P.amilis v1.0</td>\n",
       "      <td>7.570000e-48</td>\n",
       "      <td>40</td>\n",
       "      <td>238</td>\n",
       "      <td>+/+</td>\n",
       "      <td>seq.submitted 2022-08-22T18:41:21.472Z UTC</td>\n",
       "      <td>34</td>\n",
       "      <td>258</td>\n",
       "      <td>35</td>\n",
       "      <td>272</td>\n",
       "      <td>182.185</td>\n",
       "      <td>95</td>\n",
       "      <td>136</td>\n",
       "      <td>13</td>\n",
       "      <td>618</td>\n",
       "      <td>487</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Views  \\\n",
       "0    phytozome-next.jgi.doe.gov/report/protein/Vung...   \n",
       "1    phytozome-next.jgi.doe.gov/report/protein/Vung...   \n",
       "2    phytozome-next.jgi.doe.gov/report/protein/Vung...   \n",
       "3    phytozome-next.jgi.doe.gov/report/protein/Vung...   \n",
       "4    phytozome-next.jgi.doe.gov/report/protein/Vung...   \n",
       "..                                                 ...   \n",
       "192  phytozome-next.jgi.doe.gov/report/protein/Mgut...   \n",
       "193  phytozome-next.jgi.doe.gov/report/protein/Acoe...   \n",
       "194  phytozome-next.jgi.doe.gov/report/protein/Acoe...   \n",
       "195  phytozome-next.jgi.doe.gov/report/protein/Pami...   \n",
       "196  phytozome-next.jgi.doe.gov/report/protein/Pami...   \n",
       "\n",
       "                    Protein                    Species       E-value  \\\n",
       "0     VuZN016.07G092700.1.p   V.unguiculata_ZN016 v1.2  0.000000e+00   \n",
       "1      VuTZ30.07G092400.1.p    V.unguiculata_TZ30 v1.2  0.000000e+00   \n",
       "2     VuCB5-2.07G096300.1.p   V.unguiculata_CB5-2 v1.1  0.000000e+00   \n",
       "3     VuSanzi.07G093900.1.p   V.unguiculata_Sanzi v1.1  0.000000e+00   \n",
       "4    VuUCR779.07G095400.1.p  V.unguiculata_UCR779 v1.1  0.000000e+00   \n",
       "..                      ...                        ...           ...   \n",
       "192      MgNONTOL.G0505.1.p     M.guttatus_NONTOL v3.1  6.440000e-34   \n",
       "193       Aqcoe5G027700.1.p            A.coerulea v3.1  3.310000e-48   \n",
       "194       Aqcoe5G027700.1.p            A.coerulea v3.1  9.160000e-36   \n",
       "195           FUN_051248-T1              P.amilis v1.0  3.590000e-48   \n",
       "196           FUN_051248-T1              P.amilis v1.0  7.570000e-48   \n",
       "\n",
       "     % identity  Align len Strands  \\\n",
       "0            85        620     +/+   \n",
       "1            85        620     +/+   \n",
       "2            85        620     +/+   \n",
       "3            85        620     +/+   \n",
       "4            85        620     +/+   \n",
       "..          ...        ...     ...   \n",
       "192          36        222     +/+   \n",
       "193          46        184     +/+   \n",
       "194          35        243     +/+   \n",
       "195          45        190     +/+   \n",
       "196          40        238     +/+   \n",
       "\n",
       "                                       Query ID  Query from  Query to  \\\n",
       "0    seq.submitted 2022-08-22T18:41:21.472Z UTC           1       618   \n",
       "1    seq.submitted 2022-08-22T18:41:21.472Z UTC           1       618   \n",
       "2    seq.submitted 2022-08-22T18:41:21.472Z UTC           1       618   \n",
       "3    seq.submitted 2022-08-22T18:41:21.472Z UTC           1       618   \n",
       "4    seq.submitted 2022-08-22T18:41:21.472Z UTC           1       618   \n",
       "..                                          ...         ...       ...   \n",
       "192  seq.submitted 2022-08-22T18:41:21.472Z UTC          40       258   \n",
       "193  seq.submitted 2022-08-22T18:41:21.472Z UTC         430       610   \n",
       "194  seq.submitted 2022-08-22T18:41:21.472Z UTC          34       258   \n",
       "195  seq.submitted 2022-08-22T18:41:21.472Z UTC         415       601   \n",
       "196  seq.submitted 2022-08-22T18:41:21.472Z UTC          34       258   \n",
       "\n",
       "     Target from  Target to  Bitscore  # identical  Positives  Gaps  \\\n",
       "0              1        611   877.085          527        567    11   \n",
       "1              1        611   877.085          527        567    11   \n",
       "2              1        611   877.085          527        567    11   \n",
       "3              1        611   877.085          527        567    11   \n",
       "4              1        611   877.085          527        567    11   \n",
       "..           ...        ...       ...          ...        ...   ...   \n",
       "192           73        273   143.280           80        124    24   \n",
       "193          288        471   182.956           84        126     3   \n",
       "194           32        274   148.288           86        130    18   \n",
       "195          274        463   182.956           86        127     3   \n",
       "196           35        272   182.185           95        136    13   \n",
       "\n",
       "     Query len  Target len  \n",
       "0          618         612  \n",
       "1          618         612  \n",
       "2          618         612  \n",
       "3          618         612  \n",
       "4          618         612  \n",
       "..         ...         ...  \n",
       "192        618         495  \n",
       "193        618         482  \n",
       "194        618         482  \n",
       "195        618         487  \n",
       "196        618         487  \n",
       "\n",
       "[197 rows x 18 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#If there are hits, there will be an export button on phytozome blast results page, otherwise it will search for the 'no hits found' text\n",
    "export_button = WebDriverWait(driver, 1000).until(EC.any_of(EC.presence_of_element_located((By.XPATH, '/html/body/div/div/main/div/div[1]/div[1]/ul/li[1]/button')), EC.presence_of_element_located((By.XPATH, '//*[@id=\"content\"]/div/div/h2'))))\n",
    "#If no hits, create empty file, otherwise download results\n",
    "try:\n",
    "    no_hits = driver.find_element(By.XPATH, '//*[@id=\"content\"]/div/div/h2')\n",
    "    empty_data = {\n",
    "    'Views':[], \n",
    "    'Protein': [], \n",
    "    'Species': [], \n",
    "    'E-Value': [], \n",
    "    '% identity': [], \n",
    "    'Align len': [], \n",
    "    'Strands': [], \n",
    "    'Query ID': [], \n",
    "    'Query from': [],\n",
    "    'Query to': [],\n",
    "    'Target from': [],\n",
    "    'Target to': [],\n",
    "    'Bitscore': [],\n",
    "    '# identical': [],\n",
    "    'Positives': [],\n",
    "    'Gaps': [],\n",
    "    'Query len': [],\n",
    "    'Target len': []\n",
    "    }\n",
    "    phy_results = pd.DataFrame(empty_data)\n",
    "except:\n",
    "    #Export as TSV file\n",
    "    driver.execute_script(\"arguments[0].click();\", export_button)\n",
    "    path = 'C:\\\\Users\\\\Tafsia Shikdar\\\\Downloads\\\\phytozome_download\\\\'\n",
    "    list = os.listdir(path)\n",
    "    #find the file name by getting the last downloaded file\n",
    "    times = [os.path.getctime(path + file) for file in list]\n",
    "    last_file_time = max(times)\n",
    "    last_file_index = times.index(last_file_time)\n",
    "    last_file = list[last_file_index]\n",
    "    #Load into a pandas dataframe\n",
    "    phy_results = pd.read_csv(path + last_file, sep = '\\t').drop('Unnamed: 0', axis = 1)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phytozome0\n",
      "Phytozome1\n",
      "Phytozome2\n",
      "Phytozome3\n",
      "Phytozome4\n",
      "Phytozome5\n",
      "Phytozome6\n",
      "Phytozome7\n",
      "Phytozome8\n",
      "Phytozome9\n",
      "Phytozome10\n",
      "Phytozome11\n",
      "Phytozome12\n",
      "Phytozome13\n",
      "Phytozome14\n",
      "Phytozome15\n",
      "Phytozome16\n",
      "Phytozome17\n",
      "Phytozome18\n",
      "Phytozome19\n",
      "Phytozome20\n",
      "Phytozome21\n",
      "Phytozome22\n",
      "Phytozome23\n",
      "Phytozome24\n",
      "Phytozome25\n",
      "Phytozome26\n",
      "Phytozome27\n",
      "Phytozome28\n",
      "Phytozome29\n",
      "Phytozome30\n",
      "Phytozome31\n",
      "Phytozome32\n",
      "Phytozome33\n",
      "Phytozome34\n",
      "Phytozome35\n",
      "Phytozome36\n",
      "Phytozome37\n",
      "Phytozome38\n",
      "Phytozome39\n",
      "Phytozome40\n",
      "Phytozome41\n",
      "Phytozome42\n",
      "Phytozome43\n",
      "Phytozome44\n",
      "Phytozome45\n",
      "Phytozome46\n",
      "Phytozome47\n",
      "Phytozome48\n",
      "Phytozome49\n",
      "Phytozome50\n",
      "Phytozome51\n",
      "Phytozome52\n",
      "Phytozome53\n",
      "Phytozome54\n",
      "Phytozome55\n",
      "Phytozome56\n",
      "Phytozome57\n",
      "Phytozome58\n",
      "Phytozome59\n",
      "Phytozome60\n",
      "Phytozome61\n",
      "Phytozome62\n",
      "Phytozome63\n",
      "Phytozome64\n",
      "Phytozome65\n",
      "Phytozome66\n",
      "Phytozome67\n",
      "Phytozome68\n",
      "Phytozome69\n",
      "Phytozome70\n",
      "Phytozome71\n",
      "Phytozome72\n",
      "Phytozome73\n",
      "Phytozome74\n",
      "Phytozome75\n",
      "Phytozome76\n",
      "Phytozome77\n",
      "Phytozome78\n",
      "Phytozome79\n",
      "Phytozome80\n",
      "Phytozome81\n",
      "Phytozome82\n",
      "Phytozome83\n",
      "Phytozome84\n",
      "Phytozome85\n",
      "Phytozome86\n",
      "Phytozome87\n",
      "Phytozome88\n",
      "Phytozome89\n",
      "Phytozome90\n",
      "Phytozome91\n",
      "Phytozome92\n",
      "Phytozome93\n",
      "Phytozome94\n",
      "Phytozome95\n",
      "Phytozome96\n",
      "Phytozome97\n",
      "Phytozome98\n",
      "Phytozome99\n",
      "Phytozome100\n",
      "Phytozome101\n",
      "Phytozome102\n",
      "Phytozome103\n",
      "Phytozome104\n",
      "Phytozome105\n",
      "Phytozome106\n",
      "Phytozome107\n",
      "Phytozome108\n",
      "Phytozome109\n",
      "Phytozome110\n",
      "Phytozome111\n",
      "Phytozome112\n",
      "Phytozome113\n",
      "Phytozome114\n",
      "Phytozome115\n",
      "Phytozome116\n",
      "Phytozome117\n",
      "Phytozome118\n",
      "Phytozome119\n",
      "Phytozome120\n",
      "Phytozome121\n",
      "Phytozome122\n",
      "Phytozome123\n",
      "Phytozome124\n",
      "Phytozome125\n",
      "Phytozome126\n",
      "Phytozome127\n",
      "Phytozome128\n",
      "Phytozome129\n",
      "Phytozome130\n",
      "Phytozome131\n",
      "Phytozome132\n",
      "Phytozome133\n",
      "Phytozome134\n",
      "Phytozome135\n",
      "Phytozome136\n",
      "Phytozome137\n",
      "Phytozome138\n",
      "Phytozome139\n",
      "Phytozome140\n",
      "Phytozome141\n",
      "Phytozome142\n",
      "Phytozome143\n",
      "Phytozome144\n",
      "Phytozome145\n",
      "Phytozome146\n",
      "Phytozome147\n",
      "Phytozome148\n",
      "Phytozome149\n",
      "Phytozome150\n",
      "Phytozome151\n",
      "Phytozome152\n",
      "Phytozome153\n",
      "Phytozome154\n",
      "Phytozome155\n",
      "Phytozome156\n",
      "Phytozome157\n",
      "Phytozome158\n",
      "Phytozome159\n",
      "Phytozome160\n",
      "Phytozome161\n",
      "Phytozome162\n",
      "Phytozome163\n",
      "Phytozome164\n",
      "Phytozome165\n",
      "Phytozome166\n",
      "Phytozome167\n",
      "Phytozome168\n",
      "Phytozome169\n",
      "Phytozome170\n",
      "Phytozome171\n",
      "Phytozome172\n",
      "Phytozome173\n",
      "Phytozome174\n",
      "Phytozome175\n",
      "Phytozome176\n",
      "Phytozome177\n",
      "Phytozome178\n",
      "Phytozome179\n",
      "Phytozome180\n",
      "Phytozome181\n",
      "Phytozome182\n",
      "Phytozome183\n",
      "Phytozome184\n",
      "Phytozome185\n",
      "Phytozome186\n",
      "Phytozome187\n",
      "Phytozome188\n",
      "Phytozome189\n",
      "Phytozome190\n",
      "Phytozome191\n",
      "Phytozome192\n",
      "Phytozome193\n",
      "Phytozome194\n",
      "Phytozome195\n",
      "Phytozome196\n"
     ]
    }
   ],
   "source": [
    "#Manually linking exported file columns to desired column names\n",
    "phy_cols_to_keys = {\n",
    "    'Link':'Views',  \n",
    "    'Protein Name': 'Protein', \n",
    "    'Organism': 'Species', \n",
    "    'Gene Name': 'Protein', \n",
    "    'Length': 'Target len',\n",
    "    'Percent Identity': '% identity',\n",
    "}\n",
    "\n",
    "num_phy_results = len(phy_results.index)\n",
    "print('Total Phytozome Results:' + str(num_phy_results))\n",
    "#goes through each phytozome hit to get the desired data\n",
    "for row in range(num_phy_results):\n",
    "    #progress update\n",
    "    print('Phytozome' + str(row))\n",
    "    data['Website'].append('Phytozome')\n",
    "    data['Entry'].append('No Data Found')\n",
    "    data['Entry Name'].append('No Data Found')\n",
    "    for phy_col in phy_cols_to_keys:\n",
    "        data[phy_col].append(phy_results.loc[row][phy_cols_to_keys[phy_col]])\n",
    "    #Peptide sequence isn't in downloaded file, so need to find on individual webpages\n",
    "    #Open individul protein webpage\n",
    "    driver.get('https://' + phy_results.loc[row]['Views'])\n",
    "    #Click button to reveal peptide seq\n",
    "    peptide_seq_reveal = WebDriverWait(driver, 1000).until(EC.presence_of_element_located((By.XPATH, '/html/body/div/div/main/div/div[4]/div[5]/label')))\n",
    "    driver.execute_script(\"arguments[0].click();\", peptide_seq_reveal)\n",
    "    #Seq split into multiple elements on web page, so get all elements and join to make one sequence, without the * at the end\n",
    "    peptide_seq_els = driver.find_elements(By.CLASS_NAME, 'cds')\n",
    "    peptide_seq = ''.join([e.text for e in peptide_seq_els])\n",
    "    final_seq = peptide_seq[:-1]\n",
    "    #Add seq to dataframe\n",
    "    data['Sequence'].append(final_seq)\n",
    "    #Split chain according to algorithm\n",
    "    split_chain_func(final_seq)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating final CSV file of data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pd.DataFrame(data)\n",
    "#Sort dataframe according to highest percentage identity\n",
    "df = df.sort_values('Percent Identity', ascending=False)\n",
    "#Renumbering after resorting\n",
    "df = df.reset_index(drop=True)\n",
    "#Name file (if name not given, name by first 5 AAs of seq blasted)\n",
    "if title != 'Protein Search':\n",
    "    name = title\n",
    "else:\n",
    "    name = params['sequence'][:5]\n",
    "#Export dataframe to CSV and save\n",
    "df.to_csv(r'C:\\Users\\Tafsia Shikdar\\Desktop\\\\' + name + '.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.4 ('blast')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "dbdf5c8dbf87b3169415e0acf39613114f5f150d38cc301bbc68c0e106389f60"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
